addons/googledrive/models.py

```diff
@@ -0,0 +1,243 @@
+
+class GoogleDriveFile(GoogleDriveFileNode, File):
+    pass
+
+
+class GoogleDriveProvider(ExternalProvider):
+    name = 'Google Drive'
+    short_name = 'googledrive'
+
+    client_id = drive_settings.CLIENT_ID
+    client_secret = drive_settings.CLIENT_SECRET
+
+    auth_url_base = '{}{}'.format(drive_settings.OAUTH_BASE_URL, 'auth?access_type=offline&approval_prompt=force')
+    callback_url = '{}{}'.format(drive_settings.API_BASE_URL, 'oauth2/v3/token')
+    auto_refresh_url = callback_url
+    refresh_time = drive_settings.REFRESH_TIME
+    expiry_time = drive_settings.EXPIRY_TIME
+
+    default_scopes = drive_settings.OAUTH_SCOPE
+    _auth_client = GoogleAuthClient()
+    _drive_client = GoogleDriveClient()
+
+    def handle_callback(self, response):
+        client = self._auth_client
+        info = client.userinfo(response['access_token'])
+        return {
+            'provider_id': info['sub'],
+            'display_name': info['name'],
+            'profile_url': info.get('profile', None)
+        }
+
+    def fetch_access_token(self, force_refresh=False):
+        self.refresh_oauth_key(force=force_refresh)
+        return self.account.oauth_key
+
+
+class UserSettings(BaseStorageAddon, BaseOAuthUserSettings):
+    oauth_provider = GoogleDriveProvider
+    serializer = GoogleDriveSerializer
+
+
+class NodeSettings(BaseStorageAddon, BaseOAuthNodeSettings):
+    oauth_provider = GoogleDriveProvider
+    provider_name = 'googledrive'
+
+    folder_id = models.TextField(default=None, null=True, blank=True)
+    folder_path = models.TextField(null=True, blank=True)
+    serializer = GoogleDriveSerializer
+    user_settings = models.ForeignKey(UserSettings, null=True, blank=True)
+
+    _api = None
+
+    @property
+    def api(self):
+        """Authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = GoogleDriveProvider(self.external_account)
+        return self._api
+
+    @property
+    def complete(self):
+        return bool(self.has_auth and self.user_settings.verify_oauth_access(
+            node=self.owner,
+            external_account=self.external_account,
+            metadata={'folder': self.folder_id}
+        ))
+
+    @property
+    def folder_name(self):
+        if not self.folder_id:
+            return None
+
+        if self.folder_path != '/':
+            # `urllib` does not properly handle unicode.
+            # encode input to `str`, decode output back to `unicode`
+            return urllib.unquote(os.path.split(self.folder_path)[1].encode('utf-8')).decode('utf-8')
+        else:
+            return '/ (Full Google Drive)'
+
+    def fetch_folder_name(self):
+        return self.folder_name
+
+    def clear_settings(self):
+        self.folder_id = None
+        self.folder_path = None
+
+    def get_folders(self, **kwargs):
+        node = self.owner
+
+        #  Defaults exist when called by the API, but are `None`
+        path = kwargs.get('path') or ''
```

Comment:
No. The API can add `path=None` to kwargs, causing `.get('path', '')` to return `None`, which it shouldn't.